src/memory/AmdFlash.cc

```diff
@@ -455,6 +455,7 @@ bool AmdFlash::isWritable(const Sector& sector) const
 uint8_t AmdFlash::read(size_t address)
 {
 	const uint8_t value = peek(address);
+	address %= size();
```

Comment:
In `peek()` it also masks, so if I do it here as well then it masks twice. I figured I’d place it a line below in case at some point in the future something is added to `read()` after the `peek()` call which also uses the address, as a reminder of sorts. At the moment indeed it’s not used so never actually compiled in.

But it’s a bit strange I agree, also what if I want to do something with address before the call to `peek()`. The cost of the `&` operation is negligible, so I’ll just move it to the front. I would like to keep it present in all the public entry points.

> At first sight it may seem that functions with a wide contract are safer, and lead to more correct programs. But I think the opposite is true.

I definitely agree with you. And ideally it’d be constrained through the type system, the robustness of `assert()` is somewhat lacking in my opinion since it’s compiled out.

Here I used the phrase “defensive programming” to mean that it avoids a class of potential issues if the narrow contract is not sufficiently enforced elsewhere in the code. But it’s not the primary motivation of the change.